Sources/SourceKitLSP/Swift/CodeActions/UpdateDocumentation.swift

```diff
@@ -0,0 +1,138 @@
+//===----------------------------------------------------------------------===//
+//
+// This source file is part of the Swift.org open source project
+//
+// Copyright (c) 2014 - 2024 Apple Inc. and the Swift project authors
+// Licensed under Apache License v2.0 with Runtime Library Exception
+//
+// See https://swift.org/LICENSE.txt for license information
+// See https://swift.org/CONTRIBUTORS.txt for the list of Swift project authors
+//
+//===----------------------------------------------------------------------===//
+
+import SwiftParser
+import SwiftRefactor
+import SwiftSyntax
+
+/// Insert a documentation template associated with a function.
+///
+/// ## Before
+///
+/// ```swift
+/// static func refactor(syntax: DeclSyntax, in context: Void) -> DeclSyntax? {}
+/// ```
+///
+/// ## After
+///
+/// ```swift
+/// ///
+/// /// - Parameters:
+/// ///   - syntax:
+/// ///   - context:
+/// /// - Returns:
+/// static func refactor(syntax: DeclSyntax, in context: Void) -> DeclSyntax? {}
+/// ```
+@_spi(Testing)
+public struct UpdateDocumentation: SyntaxRefactoringProvider {
+  @_spi(Testing)
+  public static func refactor(syntax: DeclSyntax, in context: Void) -> DeclSyntax? {
+    let hasDocumentation = syntax.leadingTrivia.contains(where: { trivia in
+      switch trivia {
+      case .blockComment(_), .docBlockComment(_), .lineComment(_), .docLineComment(_):
+        return true
+      default:
+        return false
+      }
+    })
+
+    guard !hasDocumentation else {
+      return nil
+    }
+
+    var content: [TriviaPiece] = [
+      .docLineComment("/// A description"),
+      .newlines(1),
+    ]
+
+    if let parameters = syntax.parameters?.parameters {
+      if parameters.count == 1, let firstParam = parameters.first {
+        let paramToken = (firstParam.secondName ?? firstParam.firstName).trimmed
+        content.append(.docLineComment("/// - Parameter \(paramToken):"))
+        content.append(.newlines(1))
+      } else {
+        content.append(.docLineComment("/// - Parameters:"))
+        content.append(.newlines(1))
+        content.append(
+          contentsOf: parameters.map({ param in
+            [
+              .docLineComment("///   - \((param.secondName ?? param.firstName ).trimmed):"),
+              .newlines(1),
+            ]
+          }).joined()
+        )
+        content.append(.docLineComment("///"))
+        content.append(.newlines(1))
+      }
+    }
+
+    if syntax.throwsKeyword != nil {
+      content.append(.docLineComment("/// - Throws:"))
+      content.append(.newlines(1))
+    }
+
+    if syntax.returnType != nil {
+      content.append(.docLineComment("/// - Returns:"))
+      content.append(.newlines(1))
+    }
+
+    return syntax.with(
+      \.leadingTrivia,
+      syntax.leadingTrivia + Trivia(pieces: content)
+    )
```

Comment:
I think using an `EditRefactoringProvider` that only inserts the comment would be better. Once we have range-based formatting in swift-format, I want to run the formatter on the text inserted by a refactoring / code action and with the current implementation, this would effectively re-format the entire function body.

Reply:
Yeah, I'm going to treat this as an insertion of the documentation text. It's much nicer to use `EditRefactoringProvider` for these.